src/programs/bamvalidate.1
.TH BAMVALIDATE 1 "October 2014" BIOBAMBAM
.SH NAME
bamvalidate - validate BAM file
.SH SYNOPSIS
.PP
.B bamvalidate
[options]
.SH DESCRIPTION
bamvalidate reads a BAM, SAM or CRAM file and checks the contained
alignments for validity. If all the alignments in the source file pass
validation, then the program exits with code 0, otherwise it exits with a
non-zero code. By default no data is produced on the standard output
channel, if the passthrough parameter is set to 1, then the alignments are
recoded to standard output according to the given output parameters.
.PP
The following key=value pairs can be given:
.PP
.B verbose=<0|1>:
Valid values are
.IP 1:
print statistics on the standard error channel at the end of a successful run
.IP 0:
do not print statistics
.PP
.B passthrough=<0|1>:
recode alignments to standard output if passthrough=1. Default is
passthrough=0.
.PP
.B level=<-1|0|1|9|11>:
set compression level of the output BAM file. Valid
values are
.IP -1:
zlib/gzip default compression level
.IP 0:
uncompressed
.IP 1:
zlib/gzip level 1 (fast) compression
.IP 9:
zlib/gzip level 9 (best) compression
.P
If libmaus has been compiled with support for igzip (see
https://software.intel.com/en-us/articles/igzip-a-high-performance-deflate-compressor-with-optimizations-for-genomic-data)
then an additional valid value is
.IP 11:
igzip compression
.PP
.B tmpfile=<filename>:
set the prefix for temporary file names
.PP
.B md5=<0|1>:
md5 checksum creation for output file. This option can only be given if
outputformat=bam. Then valid values are
.IP 0:
do not compute checksum. This is the default.
.IP 1:
compute checksum. If the md5filename key is set, then the checksum is
written to the given file. If md5filename is unset, then no checksum will be computed.
.PP
.B md5filename
file name for md5 checksum if md5=1.
.PP
.B index=<0|1>:
compute BAM index for output file. This option can only be given if
outputformat=bam. Then valid values are
.IP 0:
do not compute BAM index. This is the default.
.IP 1:
compute BAM index. If the indexfilename key is set, then the BAM index is
written to the given file. If indexfilename is unset, then no BAM index will be computed.
.PP
.B indexfilename
file name for output BAM index if index=1.
.PP
.B inputformat=<bam>: input file format.
All versions of bamsort come with support for the BAM input format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit)
.PP
.B outputformat=<bam>: output file format.
All versions of bamsort come with support for the BAM output format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit). This format is not advisable for data sorted by query name.
.PP
.B I=<[stdin]>: 
input filename, standard input if unset.
.PP
.B O=<[stdout]>: 
output filename, standard output if unset.
.PP
.B inputthreads=<[1]>:
input helper threads, only valid for inputformat=bam.
.PP
.B outputthreads=<[1]>:
output helper threads, only valid for outputformat=bam.
.PP
.B reference=<[]>:
reference FastA file for inputformat=cram and outputformat=cram. An index file (.fai) is required. 
.PP
.B range=<>:
input range to be processed. This option is only valid if the input is a coordinate sorted and indexed BAM file
.PP
.B basequalhist=<[0]>:
compute base quality histogram and output this histogram on the standard
error channel after a successful validation run. The histogram lines
are prefixed with [H]. There is one line for each occurring base quality value
and two lines in the end specifying the minimum and maximum occurring
quality value. The lines contain tabulator symbol separated columns.
The occurring base quality lines have 5 columns after the [H] column. These
designate the numerical quality level, the ASCII character representation of
this level (e.g. ! for quality 0 or I for 40), the absolute number of
occurrences of the value, the fraction of the occurrences of the value (the
absolute number for this value divided by the sum of the occurrences over all
values) and the cumulative fraction for all values up to and including the
current value. The minimum and maximum value lines have min and max
respectively in the second column. The third and fourth column contain the
numerical value and the ASCII representation of this value. For all lines
the ASCII representation of a base quality value may be empty if the numerical quality value does
not correspond to a printable ASCII code.
.SH AUTHOR
Written by German Tischler.
.SH "REPORTING BUGS"
Report bugs to <tischler@mpi-cbg.de>
.SH COPYRIGHT
Copyright \(co 2009-2014 German Tischler, \(co 2011-2014 Genome Research Limited.
License GPLv3+: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100

biobambam2 (2.0.22) unstable; urgency=medium

  * implicitly concatenate multiple input files as input for bamsort 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 29 Oct 2015 09:59:48 +0100

biobambam2 (2.0.21) unstable; urgency=medium

  * fix FF and RR type alignments in streaming duplicate marking
  * add MC tag (mate cigar string) when duplicate marking support is enabled in bamsort

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 26 Oct 2015 17:12:04 +0100

biobambam2 (2.0.20) unstable; urgency=medium

  * fix marking of FF and RR type duplicate records in BAM files 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 23 Oct 2015 14:55:46 +0200

biobambam2 (2.0.19) unstable; urgency=medium

  * include updates by Andrew Whitwham for improving the cigar handling of ClipAdapters

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 22 Sep 2015 16:51:25 +0200

biobambam2 (2.0.18) unstable; urgency=medium

  * update to libmaus2 2.0.68 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 14 Sep 2015 14:03:00 +0200

biobambam2 (2.0.17) unstable; urgency=medium

  * allow building without requirement for SSE4 (change in libmaus2) 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 03 Sep 2015 10:39:07 +0200

biobambam2 (2.0.16) unstable; urgency=medium

  * merge Andrew Whitwham's patch for removing adapter clipping tags from secondary and supplementary alignments 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 24 Aug 2015 13:15:02 +0200

biobambam2 (2.0.15) unstable; urgency=medium

  * update to libmaus2 2.0.45 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 21 Aug 2015 10:13:18 +0200

biobambam2 (2.0.14) unstable; urgency=medium

  * update to libmaus2 2.0.42 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 09 Aug 2015 13:36:32 +0200

biobambam2 (2.0.13) unstable; urgency=medium

  * allow in memory tmp files in bammarkduplicates, bammarkduplicates2 and bamtofastq 

 -- German Tischler <gt1@sanger.ac.uk>  Sat, 08 Aug 2015 15:03:19 +0200

biobambam2 (2.0.12) unstable; urgency=medium

  * update to libmaus2 2.0.40

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 07 Aug 2015 08:36:53 +0200

biobambam2 (2.0.11) unstable; urgency=medium

  * update to libmaus2 2.0.39 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 06 Aug 2015 18:23:03 +0200

biobambam2 (2.0.10) unstable; urgency=medium

  * include Andrew Whitwham's (streaming) duplicate marking tag changes

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 22 Jun 2015 11:22:34 +0200

biobambam2 (2.0.9) unstable; urgency=medium

  * add daligner las to SAM/BAM/CRAM conversion tool 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 19 Jun 2015 15:45:53 +0200

biobambam2 (2.0.8) unstable; urgency=low

  * fix for random failures in bamsormadup due to missing/incomplete initialisation 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 23:49:56 +0100

biobambam2 (2.0.7) unstable; urgency=low

  * update libmaus version to 2.0.6 (fix for bam indexing) 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 11:43:45 +0100

biobambam2 (2.0.6) unstable; urgency=low

  * update libmaus version to 2.0.5 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 24 Apr 2015 09:31:27 +0100

biobambam2 (2.0.5) unstable; urgency=low

  * fix/clarify manual page for bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 15:33:14 +0100

biobambam2 (2.0.4) unstable; urgency=low

  * add option to store temp data in RAM instead of on disc in bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 13:53:51 +0100

biobambam2 (2.0.3) unstable; urgency=medium

  * update to libmaus 2.0.3 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 12:48:32 +0100

biobambam2 (2.0.2) unstable; urgency=medium

  * update to libmaus 2.0.2 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 10:36:05 +0100

biobambam2 (2.0.1) unstable; urgency=medium

  * no change version bump to check version bumping script 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 19 Apr 2015 00:15:29 +0100

biobambam2 (2.0.0) unstable; urgency=low

  * import of biobambam code

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 18 Apr 2015 12:00:00 +0100


src/programs/bamvalidate.cpp
/**
    biobambam
    Copyright (C) 2009-2014 German Tischler
    Copyright (C) 2011-2014 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include "config.h"

#include <iostream>

#include <libmaus2/bambam/BamMultiAlignmentDecoderFactory.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/bambam/BgzfDeflateOutputCallbackBamIndex.hpp>
#include <libmaus2/bambam/ProgramHeaderLineSet.hpp>
#include <libmaus2/lz/BgzfDeflateOutputCallbackMD5.hpp>
#include <libmaus2/util/ArgInfo.hpp>
#include <libmaus2/util/TempFileRemovalContainer.hpp>

#include <biobambam2/BamBamConfig.hpp>
#include <biobambam2/Licensing.hpp>

int getDefaultMD5()
{
	return 0;
}
int getDefaultIndex()
{
	return 0;
}

int getDefaultLevel()
{
	return Z_DEFAULT_COMPRESSION;
}

bool getDefaultBaseQualHist()
{
	return false;
}

bool getDefaultPassThrough()
{
	return false;
}

int getDefaultVerbose()
{
	return false;
}

char const * getDefaultInputFormat()
{
	return "bam";
}

template<bool passthrough>
int bamvalidateTemplate(::libmaus2::util::ArgInfo const & arginfo)
{
	libmaus2::timing::RealTimeClock rtc; rtc.start();	
	bool const verbose = arginfo.getValue("verbose",getDefaultVerbose());
	bool const basequalhist = arginfo.getValue("basequalhist",getDefaultBaseQualHist());
	
	libmaus2::bambam::BamAlignmentDecoderWrapper::unique_ptr_type decwrapper(
		libmaus2::bambam::BamMultiAlignmentDecoderFactory::construct(
			arginfo,false // put rank
		)
	);
	::libmaus2::bambam::BamAlignmentDecoder * ppdec = &(decwrapper->getDecoder());
	::libmaus2::bambam::BamAlignmentDecoder & dec = *ppdec;
	::libmaus2::bambam::BamHeader const & header = dec.getHeader();	
	::libmaus2::bambam::BamAlignment const & algn = dec.getAlignment();

	// add PG line to header
	std::string const upheadtext = ::libmaus2::bambam::ProgramHeaderLineSet::addProgramLine(
		header.text,
		"bamvalidate", // ID
		"bamvalidate", // PN
		arginfo.commandline, // CL
		::libmaus2::bambam::ProgramHeaderLineSet(header.text).getLastIdInChain(), // PP
		std::string(PACKAGE_VERSION) // VN			
	);
	// construct new header
	::libmaus2::bambam::BamHeader uphead(upheadtext);

	/*
	 * start index/md5 callbacks and alignment writer
	 */
	std::string md5filename;
	std::string indexfilename;

	std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > cbs;
	::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Pmd5cb;
	libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Pindex;
	libmaus2::bambam::BamBlockWriterBase::unique_ptr_type Pout;

	if ( passthrough )
	{
		std::string const tmpfilenamebase = arginfo.getValue<std::string>("tmpfile",arginfo.getDefaultTmpFileName());
		std::string const tmpfileindex = tmpfilenamebase + "_index";
		::libmaus2::util::TempFileRemovalContainer::addTempFile(tmpfileindex);
		
		if ( arginfo.getValue<unsigned int>("md5",getDefaultMD5()) )
		{
			if ( arginfo.hasArg("md5filename") &&  arginfo.getUnparsedValue("md5filename","") != "" )
				md5filename = arginfo.getUnparsedValue("md5filename","");
			else
				std::cerr << "[V] no filename for md5 given, not creating hash" << std::endl;

			if ( md5filename.size() )
			{
				::libmaus2::lz::BgzfDeflateOutputCallbackMD5::unique_ptr_type Tmd5cb(new ::libmaus2::lz::BgzfDeflateOutputCallbackMD5);
				Pmd5cb = UNIQUE_PTR_MOVE(Tmd5cb);
				cbs.push_back(Pmd5cb.get());
			}
		}
		if ( arginfo.getValue<unsigned int>("index",getDefaultIndex()) )
		{
			if ( arginfo.hasArg("indexfilename") &&  arginfo.getUnparsedValue("indexfilename","") != "" )
				indexfilename = arginfo.getUnparsedValue("indexfilename","");
			else
				std::cerr << "[V] no filename for index given, not creating index" << std::endl;

			if ( indexfilename.size() )
			{
				libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex::unique_ptr_type Tindex(new libmaus2::bambam::BgzfDeflateOutputCallbackBamIndex(tmpfileindex));
				Pindex = UNIQUE_PTR_MOVE(Tindex);
				cbs.push_back(Pindex.get());
			}
		}
		std::vector< ::libmaus2::lz::BgzfDeflateOutputCallback * > * Pcbs = 0;
		if ( cbs.size() )
			Pcbs = &cbs;


		libmaus2::bambam::BamBlockWriterBase::unique_ptr_type Tout ( 
			libmaus2::bambam::BamBlockWriterBaseFactory::construct(uphead, arginfo, Pcbs) 
		);
		Pout = UNIQUE_PTR_MOVE(Tout);
	}
	
	libmaus2::autoarray::AutoArray<char> lastvalidname(256); // max valid read name is 255 bytes
	uint64_t alsok = 0;

	::libmaus2::autoarray::AutoArray<char> qual;
	libmaus2::autoarray::AutoArray<uint64_t> H(static_cast<uint64_t>(std::numeric_limits<uint8_t>::max())+1);
	std::fill(H.begin(),H.end(),0ull);
	
	try
	{
		while ( dec.readAlignment() )
		{
			if ( passthrough )
				Pout->writeAlignment(algn);

			if ( basequalhist )
			{
				uint64_t const l = algn.getLseq();
				uint8_t const * Qc = libmaus2::bambam::BamAlignmentDecoderBase::getQual(algn.D.begin());					
				uint8_t const * const Qe = Qc + l;
				
				while ( Qc != Qe )
					H[*(Qc++)]++;
			}

			uint64_t const lname = algn.getLReadName();
			char const * name = algn.getName();
			std::copy(name,name+lname+1,lastvalidname.begin());
			
			alsok += 1;
		}
	}
	catch(std::exception const & ex)
	{
		std::cerr << "[E] name of last valid alignment was " << lastvalidname.begin() << std::endl;
		std::cerr << "[E] read " << alsok << " valid alignments" << std::endl;
		throw;
	}
	
	Pout.reset();

	if ( Pmd5cb )
	{
		Pmd5cb->saveDigestAsFile(md5filename);
	}
	if ( Pindex )
	{
		Pindex->flush(std::string(indexfilename));
	}

	if ( verbose )
		std::cerr << "[V] checked " << alsok << " alignments in " << rtc.formatTime(rtc.getElapsedSeconds()) 
			<< " (" << alsok / rtc.getElapsedSeconds() << " al/s)" << std::endl;
			
	if ( basequalhist )
	{
		uint64_t const s = std::accumulate(H.begin(),H.end(),0ull);
		
		uint64_t a = 0;		
		uint64_t minq = std::numeric_limits<uint64_t>::max();
		uint64_t maxq = 0;
		
		for ( uint64_t i = 0; i < H.size(); ++i )
			if ( H[i] )
			{
				minq = std::min(minq,i);
				maxq = std::max(maxq,i);
			
				a += H[i];
				
				std::cerr 
					<< "[H]\t" << i << "\t";
				
				if ( ( static_cast<uint64_t>(i+33) < static_cast<uint64_t>(std::numeric_limits<char>::max()) && isprint(i+33)) )
					std::cerr << static_cast<char>(i+33);
				
				std::cerr << "\t"
					<< H[i] << "\t" 
					<< (H[i] / static_cast<double>(s)) << "\t"
					<< (a / static_cast<double>(s))
					<< std::endl;
			}
			
		if ( s )
		{
			std::cerr << "[H]\tmin\t" << minq << "\t";
			if ( ( static_cast<uint64_t>(minq+33) < static_cast<uint64_t>(std::numeric_limits<char>::max()) && isprint(minq+33)) )
				std::cerr << static_cast<char>(minq+33);
			std::cerr << std::endl;
			std::cerr << "[H]\tmax\t" << maxq << "\t";
			if ( ( static_cast<uint64_t>(maxq+33) < static_cast<uint64_t>(std::numeric_limits<char>::max()) && isprint(maxq+33)) )
				std::cerr << static_cast<char>(maxq+33);
			std::cerr << std::endl;
		}
	}
	
	return EXIT_SUCCESS;
}

int bamvalidate(::libmaus2::util::ArgInfo const & arginfo)
{
	bool const passthrough = arginfo.getValue<unsigned int>("passthrough",getDefaultPassThrough());
	
	if ( passthrough )
		return bamvalidateTemplate<true>(arginfo);
	else
		return bamvalidateTemplate<false>(arginfo);
}

int main(int argc, char * argv[])
{
	try
	{
		::libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if ( 
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if ( 
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;
				
				std::vector< std::pair<std::string,std::string> > V;
			
				V.push_back ( std::pair<std::string,std::string> ( "verbose=<["+::biobambam2::Licensing::formatNumber(getDefaultVerbose())+"]>", "print stats at the end of a successfull run" ) );
				V.push_back ( std::pair<std::string,std::string> ( "basequalhist=<["+::biobambam2::Licensing::formatNumber(getDefaultBaseQualHist())+"]>", "print base quality histogram at end of a successfull run" ) );
				V.push_back ( std::pair<std::string,std::string> ( "passthrough=<["+::biobambam2::Licensing::formatNumber(getDefaultPassThrough())+"]>", "write alignments to standard output (default: do not pass through)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "level=<["+::biobambam2::Licensing::formatNumber(getDefaultLevel())+"]>", libmaus2::bambam::BamBlockWriterBaseFactory::getBamOutputLevelHelpText() ) );
				V.push_back ( std::pair<std::string,std::string> ( "tmpfile=<filename>", "prefix for temporary files, default: create files in current directory (passthrough=1, index=1 only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5=<["+::biobambam2::Licensing::formatNumber(getDefaultMD5())+"]>", "create md5 check sum (default: 0, passthrough=1 only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "md5filename=<filename>", "file name for md5 check sum" ) );
				V.push_back ( std::pair<std::string,std::string> ( "index=<["+::biobambam2::Licensing::formatNumber(getDefaultIndex())+"]>", "create BAM index (default: 0, passthrough=1 only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "indexfilename=<filename>", "file name for BAM index file" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("inputformat=<[")+getDefaultInputFormat()+"]>", std::string("input format (") + libmaus2::bambam::BamMultiAlignmentDecoderFactory::getValidInputFormats() + ")" ) );
				V.push_back ( std::pair<std::string,std::string> ( std::string("outputformat=<[")+libmaus2::bambam::BamBlockWriterBaseFactory::getDefaultOutputFormat()+"]>", std::string("output format (") + libmaus2::bambam::BamBlockWriterBaseFactory::getValidOutputFormats() + ", passthrough=1 only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "I=<[stdin]>", "input filename (standard input if unset)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "inputthreads=<[1]>", "input helper threads (for inputformat=bam only, default: 1)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "reference=<>", "reference FastA (.fai file required, for cram i/o only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "range=<>", "coordinate range to be processed (for coordinate sorted indexed BAM input only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "outputthreads=<[1]>", "output helper threads (for outputformat=bam only, default: 1, passthrough=1 only)" ) );
				V.push_back ( std::pair<std::string,std::string> ( "O=<[stdout]>", "output filename (standard output if unset, passthrough=1 only)" ) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;
				return EXIT_SUCCESS;
			}


		return bamvalidate(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}

}
